### PR TITLE
Fix/3520

### LINKED ISSUE #3520: 2.1: Affirmation logic in burnchains::db panics due to burnchain fork
The assertion at https://github.com/stacks-network/stacks-blockchain/blob/develop/src/burnchains/db.rs#L732 can cause a panic in certain burnchain forks. This was seen in Hiro's 2.1 testnet node. It was running an old build, but this bug is present in `develop` as well.

Relevant portion of the stack trace:

```
Process abort due to thread panic: panicked at 'assertion failed: `(left == right)`   left: `42`,  right: `44`', src/burnchains/db.rs:729:9

Panic backtrace:    0: stacks_node::main::{{closure}}
              at src/testnet/stacks-node/src/main.rs:59:18
    1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
...
    7: core::panicking::assert_failed
              at rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:203:5
    8: blockstack_lib::burnchains::db::BurnchainDBTransaction::make_reward_phase_affirmation_map
              at src/src/burnchains/db.rs:729:9
    9: blockstack_lib::burnchains::db::BurnchainDBTransaction::update_block_descendancy
              at src/src/burnchains/db.rs:469:21
    10: blockstack_lib::burnchains::db::BurnchainDBTransaction::store_blockstack_ops
              at src/src/burnchains/db.rs:901:9
```

After a little digging it seems this was caused by a burnchain fork at height 2417545

```sql
sqlite> select block_hash from burnchain_db_block_headers where block_height = 2417545;
000000000000000b50b99180deeea0f92a379cd65414df4af02dc74707e7306d
000000003126084450b3872ec609055f986c48a300accc230a8340923c86c29b
```

There's two blocks there, and each of them include 2 stacks operations:

```sql
sqlite> SELECT txid FROM burnchain_db_block_ops WHERE block_hash = "000000003126084450b3872ec609055f986c48a300accc230a8340923c86c29b";
c0814b6585386f3b100fb8870c2cbeb3ba27f0fbfe7b8b183d146625a2c5363c
042e1a4b49926c963f6661dcf575308e51e969166a35ff9e0364677dc3dc23f6
sqlite> SELECT txid FROM burnchain_db_block_ops WHERE block_hash = "000000000000000b50b99180deeea0f92a379cd65414df4af02dc74707e7306d";
c0814b6585386f3b100fb8870c2cbeb3ba27f0fbfe7b8b183d146625a2c5363c
042e1a4b49926c963f6661dcf575308e51e969166a35ff9e0364677dc3dc23f6
```

However, those txids have different vtx indexes in the two forks -- the relevant difference is that in one fork it is 44 and the other it is 42.

So how does `make_reward_phase_affirmation_map()` end up with an arg from one fork and an arg from the other fork? In `update_block_descendancy`, `get_commit_at` is invoked, which invokes `get_commit_in_block_at`. This method selects the txid from `block_commit_metadata` table, and then uses the txid to lookup in the `burnchain_db_block_ops` table. However, txid by itself cannot differentiate between the two versions of the transaction that appears in the block ops table (in both forks, the txid is the same, but the vtxindex is different).

One "quick fix" would be to add the burnchain header hash to that last lookup, but I'm wondering if this is surfacing a different issue for the affirmation logic in burnchains::db.

### PR DESCRIPTION
**Pull request by jcnelson**
This fixes #3520 by requiring that any query against the burnchain DB happens with respect to a burnchain block header hash.  This is usually obtained from an instance of `BurnchainHeaderReader`, which is used to select previously-stored burnchain operations that are on the burnchain fork represented by that instance.  In the few cases where a `BurnchainHeaderReader` is not used, a "known canonical" burnchain block hash is supplied in order to constrain the query to only returning data for the burnchain fork that contains this canonical burnchain block hash.

Because this PR touches many parts of the burnchain DB and the chains coordinator, I'd recommend re-syncing this from genesis on testnet and mainnet to verify that it works correctly.

### CODE REVIEW

File: stacks-common/src/util/uint.rs [hunk 1]
```diff
@@ -480,13 +480,13 @@ impl Uint256 {
     #[inline]
     pub fn increment(&mut self) {
         let &mut Uint256(ref mut arr) = self;
-        arr[0] += 1;
+        arr[0] = arr[0].wrapping_add(1);
```
[the remark this answers is not in the document]
Comment (kantai) [reply stacks-common/src/util/uint.rs:483]: Is there a test case for this behavior?

File: src/burnchains/tests/db.rs [hunk 2]
```diff
@@ -579,20 +591,38 @@ fn test_get_commit_at() {
...
-    assert!(cmt.is_none());
+    let cmt = BurnchainDB::get_commit_at(&burnchain_db.conn(), &headers, 5, 0)
+        .unwrap()
+        .unwrap();
+    assert_eq!(cmt, cmts[4]);
+
+    let cmt = BurnchainDB::get_commit_at(&burnchain_db.conn(), &fork_headers, 5, 1)
+        .unwrap()
+        .unwrap();
+    assert_eq!(cmt, fork_cmt);
```
Comment (kantai) [inline src/burnchains/tests/db.rs:625]: Can you add a brief comment here that this test covers #3520 and a description of what it's checking?

**Review by kantai — COMMENTED** [review]
These changes look good to me, but the changes to the Uint256 need test cases

**Review by obycode — APPROVED** [review]
LGTM!

File: src/burnchains/tests/db.rs (hunk 2 above)
Comment (jcnelson) [reply src/burnchains/tests/db.rs]: Better: I added a distinct test for it

File: stacks-common/src/util/uint.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcnelson) [reply stacks-common/src/util/uint.rs:483]: There is now.  Also, this is what upstream does now.

File: src/burnchains/tests/db.rs (hunk 2 above)
Comment (jcnelson) [reply src/burnchains/tests/db.rs]: The test: https://github.com/stacks-network/stacks-blockchain/pull/3524/files#diff-53d26ed8ff095df944d5c90e1c20a7792aaa5032d510a1ffc93bd584e0f58aa5R833

File: stacks-common/src/util/uint.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcnelson) [reply stacks-common/src/util/uint.rs:483]: The test: https://github.com/stacks-network/stacks-blockchain/pull/3524/files#diff-fd17390f3effc142fd313eed1a4f60794904db27dac2d510178d1ac81699ad80R756

**Review by kantai — APPROVED** [review]
LGTM!

**Pull request merged**